File: tools/ts-reference/engine/prestige.ts
import { Decimal, D } from './bignum';
import { GasId, GAS_LIST } from './gases';
import { ResourceId } from './resources';
import { GasTotals } from './gameState';
import { PRESTIGE, PRESTIGE_CURRENCY_NAME, PRESTIGE_CURRENCY_SHORT } from './constants';

export { PRESTIGE_CURRENCY_NAME, PRESTIGE_CURRENCY_SHORT };

export interface PrestigeUpgradeEffect {
  /** Multiplies all resource + gas production, stacking per level. */
  globalProductionMultiplier?: number;
  /** Multiplies Research income specifically, stacking per level. */
  researchMultiplier?: number;
  /** Multiplies production of one specific gas, stacking per level. */
  gasProductionMultiplier?: { gas: GasId; multiplier: number };
  /** Multiplies ALL greenhouse-gas production at once (distinct from resource currencies). */
  allGasProductionMultiplier?: number;
  /** Fractional discount applied to every technology's cost (0.1 = 10% cheaper), stacking additively across levels, capped at 0.9. */
  techCostDiscount?: number;
  /** Multiplies the base offline-progress cap, stacking per level. */
  offlineCapMultiplier?: number;
  /** Technologies auto-granted (owned = 1) at the start of every future run. */
  grantsStartingTechIds?: string[];
  /** Extra owned units of a generator granted at the start of every future run, per level. */
  startingGenerators?: Record<string, number>;
  /** Flat resources granted at the start of every future run. */
  startingResources?: Partial<Record<ResourceId, number>>;
}

export interface PrestigeUpgrade {
  id: string;
  name: string;
  description: string;
  icon: string;
  baseCost: number; // in Earth Points
  costGrowth: number; // per level; 1 for one-time upgrades
  maxLevel: number;
  effect: PrestigeUpgradeEffect;
}

export const PRESTIGE_UPGRADES: PrestigeUpgrade[] = [
  {
    id: 'atmospheric_momentum',
    name: 'Atmospheric Momentum',
    description: 'Each civilization remembers a little more than the last. +10% production per level.',
    icon: '🌬️',
    baseCost: 100,
    costGrowth: 2.2,
    maxLevel: Infinity,
    effect: { globalProductionMultiplier: 1.1 },
  },
  {
    id: 'industrial_memory',
    name: 'Industrial Memory',
    description: 'Start every future Earth already knowing the Steam Engine.',
    icon: '⚙️',
    baseCost: 1_000,
    costGrowth: 1,
    maxLevel: 1,
    effect: { grantsStartingTechIds: ['steam_engine'] },
  },
  {
    id: 'rapid_research',
    name: 'Rapid Research',
    description: 'Institutional knowledge survives the reset. +50% Research per level.',
    icon: '📚',
    baseCost: 10_000,
    costGrowth: 3,
    maxLevel: 5,
    effect: { researchMultiplier: 1.5 },
  },
  {
    id: 'institutional_memory',
    name: 'Institutional Memory',
    description: 'Every civilization inherits the last one\'s blueprints. Every building you own produces +40% more, per level.',
    icon: '🏛️',
    baseCost: 150_000,
    costGrowth: 4,
    maxLevel: 6,
    effect: { globalProductionMultiplier: 1.4 },
  },
  {
    id: 'civilizational_acceleration',
    name: 'Civilizational Acceleration',
    description: 'Blueprints from past civilizations make every technology 20% cheaper per level.',
    icon: '📐',
    baseCost: 1_000_000,
    costGrowth: 4,
    maxLevel: 4,
    effect: { techCostDiscount: 0.2 },
  },
  {
    id: 'anthropocene_mastery',
    name: 'Anthropocene Mastery',
    description: 'You\'ve done this before. ×10 all greenhouse-gas production.',
    icon: '🌋',
    baseCost: 25_000_000,
    costGrowth: 1,
    maxLevel: 1,
    effect: { allGasProductionMultiplier: 10 },
  },
  {
    id: 'eternal_flame',
    name: 'Eternal Flame',
    description: 'An ember carried between worlds. Every future Earth starts with its fires already burning: +2 Natural Fire per level.',
    icon: '🔥',
    baseCost: 50,
    costGrowth: 1.8,
    maxLevel: 10,
    effect: { startingGenerators: { natural_fire: 2 } },
  },
  {
    id: 'extended_endurance',
    name: 'Extended Endurance',
    description: 'Your civilizations keep running longer without you watching. +100% offline progress cap per level.',
    icon: '⏳',
    baseCost: 5_000,
    costGrowth: 3.5,
    maxLevel: 3,
    effect: { offlineCapMultiplier: 2 },
  },
  {
    id: 'head_start',
    name: 'Head Start',
    description: 'Begin each Earth with a stockpile of Energy, skipping the slowest opening minutes.',
    icon: '🚩',
    baseCost: 20_000,
    costGrowth: 1,
    maxLevel: 1,
    effect: { startingResources: { energy: 1000 } },
  },
  {
    id: 'automated_industry',
    name: 'Automated Industry',
    description: 'Robots keep the lines running while you sleep. +100% offline progress cap, and every future Earth starts with 25 Natural Fires already lit.',
    icon: '🤖',
    baseCost: 200_000,
    costGrowth: 1,
    maxLevel: 1,
    effect: { offlineCapMultiplier: 2, startingGenerators: { natural_fire: 25 } },
  },
  {
    id: 'deep_foundations',
    name: 'Deep Foundations',
    description: 'Start every future Earth with a Research head start, so the tree opens immediately instead of after the first hour.',
    icon: '🧱',
    baseCost: 40_000,
    costGrowth: 1,
    maxLevel: 1,
    effect: { startingResources: { research: 5_000, energy: 25_000 } },
  },
];

export const PRESTIGE_UPGRADE_BY_ID: Record<string, PrestigeUpgrade> = Object.fromEntries(
  PRESTIGE_UPGRADES.map((u) => [u.id, u]),
);

export function prestigeUpgradeCost(upgrade: PrestigeUpgrade, currentLevel: number): Decimal {
  return D(upgrade.baseCost).mul(D(upgrade.costGrowth).pow(currentLevel));
}

export interface PrestigeMultipliers {
  global: number;
  research: number;
  allGas: number;
  perGas: Partial<Record<GasId, number>>;
  techCostDiscount: number;
  offlineCapMultiplier: number;
  startingTechIds: string[];
  startingGenerators: Record<string, number>;
  startingResources: Partial<Record<ResourceId, number>>;
}

function applyEffect(result: PrestigeMultipliers, e: PrestigeUpgradeEffect, level: number): void {
  if (e.globalProductionMultiplier) result.global *= Math.pow(e.globalProductionMultiplier, level);
  if (e.researchMultiplier) result.research *= Math.pow(e.researchMultiplier, level);
  if (e.allGasProductionMultiplier) result.allGas *= Math.pow(e.allGasProductionMultiplier, level);
  if (e.gasProductionMultiplier) {
    const { gas, multiplier } = e.gasProductionMultiplier;
    result.perGas[gas] = (result.perGas[gas] ?? 1) * Math.pow(multiplier, level);
  }
  if (e.techCostDiscount) result.techCostDiscount = Math.min(0.9, result.techCostDiscount + e.techCostDiscount * level);
  if (e.offlineCapMultiplier) result.offlineCapMultiplier *= Math.pow(e.offlineCapMultiplier, level);
  if (e.grantsStartingTechIds) result.startingTechIds.push(...e.grantsStartingTechIds);
  if (e.startingGenerators) {
    for (const [techId, count] of Object.entries(e.startingGenerators)) {
      result.startingGenerators[techId] = (result.startingGenerators[techId] ?? 0) + count * level;
    }
  }
  if (e.startingResources) {
    for (const [res, amount] of Object.entries(e.startingResources)) {
      result.startingResources[res as ResourceId] = (result.startingResources[res as ResourceId] ?? 0) + (amount ?? 0);
    }
  }
}

/**
 * Aggregates every owned prestige upgrade (at every level) plus any
 * permanent challenge rewards into one multiplier bundle. Challenge
 * rewards use the same effect shape as prestige upgrades and are folded in
 * at a flat level of 1 (they don't stack with themselves — a challenge is
 * either completed once or not).
 */
export function computePrestigeMultipliers(
  upgradesOwned: Record<string, number>,
  challengeRewardEffects: PrestigeUpgradeEffect[] = [],
): PrestigeMultipliers {
  const result: PrestigeMultipliers = {
    global: 1,
    research: 1,
    allGas: 1,
    perGas: {},
    techCostDiscount: 0,
    offlineCapMultiplier: 1,
    startingTechIds: [],
    startingGenerators: {},
    startingResources: {},
  };

  for (const upgrade of PRESTIGE_UPGRADES) {
    const level = upgradesOwned[upgrade.id] ?? 0;
    if (level <= 0) continue;
    applyEffect(result, upgrade.effect, level);
  }

  for (const rewardEffect of challengeRewardEffects) {
    applyEffect(result, rewardEffect, 1);
  }

  return result;
}

export interface PrestigeGainParams {
  totalGasProducedKg: GasTotals;
  peakForcingWm2: number;
  civLevel: number;
  runDurationSeconds: number;
}

function sumAllGasKg(totals: GasTotals): Decimal {
  return GAS_LIST.reduce((sum, gas) => sum.add(totals[gas.id]), Decimal.ZERO);
}

/**
 * How much faster than the reference pace this run was, as a multiplier on
 * the payout. This is the term that makes each Earth worth more than the
 * last: prestige upgrades buy speed, and speed is what is scored.
 *
 * A run of zero (or nonsensical) length is treated as the reference pace
 * rather than as infinitely fast, so a corrupted clock cannot mint points.
 */
export function speedMultiplier(runDurationSeconds: number): number {
  if (!Number.isFinite(runDurationSeconds) || runDurationSeconds <= 0) return 1;
  const ratio = PRESTIGE.referenceRunSeconds / runDurationSeconds;
  const scaled = Math.pow(ratio, PRESTIGE.speedExponent);
  return Math.min(PRESTIGE.maxSpeedMultiplier, Math.max(PRESTIGE.minSpeedMultiplier, scaled));
}

/**
 * Converts one run's outcome into Earth Points: how much gas the civilization
 * produced, scaled by how hard it pushed the atmosphere, how far up the tech
 * tree it got, and how quickly it managed all three. Every tunable lives in
 * `constants.ts > PRESTIGE`, which also explains why each term is shaped the
 * way it is.
 */
export function calculatePrestigeGain(params: PrestigeGainParams): Decimal {
  const totalGasKg = sumAllGasKg(params.totalGasProducedKg);
  const gasScore = totalGasKg.div(PRESTIGE.baseDivisor).clampMin(0);
  const gasComponent = gasScore.pow(PRESTIGE.exponent).mul(PRESTIGE.gasWeight);

  const forcingMultiplier = 1 + PRESTIGE.forcingWeight * Math.sqrt(Math.max(0, params.peakForcingWm2));
  const civMultiplier = 1 + PRESTIGE.civLevelWeight * Math.sqrt(Math.max(0, params.civLevel));

  return gasComponent
    .mul(forcingMultiplier)
    .mul(civMultiplier)
    .mul(speedMultiplier(params.runDurationSeconds));
}

File: app/src/main/java/com/earthgame/idle/domain/prestige/Prestige.kt
package com.earthgame.idle.domain.prestige

import com.earthgame.idle.domain.engine.GameDecimal
import com.earthgame.idle.domain.engine.PRESTIGE
import com.earthgame.idle.domain.engine.gd
import com.earthgame.idle.domain.model.GasAmounts
import com.earthgame.idle.domain.model.GasId
import com.earthgame.idle.domain.model.ResourceId
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * What one prestige upgrade level does. Challenge rewards use the same shape,
 * folded in at a flat level of 1 — a challenge is either completed once or not.
 */
data class PrestigeUpgradeEffect(
    /** Multiplies all resource and gas production, stacking per level. */
    val globalProductionMultiplier: Double? = null,
    /** Multiplies Research income specifically, stacking per level. */
    val researchMultiplier: Double? = null,
    /** Multiplies production of one specific gas, stacking per level. */
    val gasProductionMultiplier: Pair<GasId, Double>? = null,
    /** Multiplies ALL greenhouse-gas production at once (distinct from resource currencies). */
    val allGasProductionMultiplier: Double? = null,
    /** Fractional discount on every technology cost, stacking additively across levels, capped at 0.9. */
    val techCostDiscount: Double? = null,
    /** Multiplies the base offline-progress cap, stacking per level. */
    val offlineCapMultiplier: Double? = null,
    /** Technologies auto-granted (owned = 1) at the start of every future run. */
    val grantsStartingTechIds: List<String> = emptyList(),
    /** Extra owned units of a generator granted at the start of every future run, per level. */
    val startingGenerators: Map<String, Int> = emptyMap(),
    /** Flat resources granted at the start of every future run. */
    val startingResources: Map<ResourceId, Double> = emptyMap(),
)

data class PrestigeUpgrade(
    val id: String,
    val displayName: String,
    val description: String,
    val icon: String,
    /** In Earth Points. */
    val baseCost: Double,
    /** Per level; 1 for one-time upgrades. */
    val costGrowth: Double,
    val maxLevel: Int,
    val effect: PrestigeUpgradeEffect,
) {
    val isRepeatable: Boolean get() = maxLevel > 1

    companion object {
        /** No ceiling. Cost growth makes anything past a few dozen levels unaffordable anyway. */
        const val UNLIMITED_LEVELS = Int.MAX_VALUE
    }
}

val PRESTIGE_UPGRADES: List<PrestigeUpgrade> = listOf(
    PrestigeUpgrade(
        id = "atmospheric_momentum",
        displayName = "Atmospheric Momentum",
        description = "Each civilization remembers a little more than the last. +10% production per level.",
        icon = "🌬️",
        baseCost = 100.0,
        costGrowth = 2.2,
        maxLevel = PrestigeUpgrade.UNLIMITED_LEVELS,
        effect = PrestigeUpgradeEffect(globalProductionMultiplier = 1.1),
    ),
    PrestigeUpgrade(
        id = "industrial_memory",
        displayName = "Industrial Memory",
        description = "Start every future Earth already knowing the Steam Engine.",
        icon = "⚙️",
        baseCost = 1_000.0,
        costGrowth = 1.0,
        maxLevel = 1,
        effect = PrestigeUpgradeEffect(grantsStartingTechIds = listOf("steam_engine")),
    ),
    PrestigeUpgrade(
        id = "rapid_research",
        displayName = "Rapid Research",
        description = "Institutional knowledge survives the reset. +50% Research per level.",
        icon = "📚",
        baseCost = 10_000.0,
        costGrowth = 3.0,
        maxLevel = 5,
        effect = PrestigeUpgradeEffect(researchMultiplier = 1.5),
    ),
    PrestigeUpgrade(
        id = "institutional_memory",
        displayName = "Institutional Memory",
        description = "Every civilization inherits the last one's blueprints. Every building you own produces +40% more, per level.",
        icon = "🏛️",
        baseCost = 150_000.0,
        costGrowth = 4.0,
        maxLevel = 6,
        effect = PrestigeUpgradeEffect(globalProductionMultiplier = 1.4),
    ),
    PrestigeUpgrade(
        id = "civilizational_acceleration",
        displayName = "Civilizational Acceleration",
        description = "Blueprints from past civilizations make every technology 20% cheaper per level.",
        icon = "📐",
        baseCost = 1_000_000.0,
        costGrowth = 4.0,
        maxLevel = 4,
        effect = PrestigeUpgradeEffect(techCostDiscount = 0.2),
    ),
    PrestigeUpgrade(
        id = "anthropocene_mastery",
        displayName = "Anthropocene Mastery",
        description = "You've done this before. ×10 all greenhouse-gas production.",
        icon = "🌋",
        baseCost = 25_000_000.0,
        costGrowth = 1.0,
        maxLevel = 1,
        effect = PrestigeUpgradeEffect(allGasProductionMultiplier = 10.0),
    ),
    PrestigeUpgrade(
        id = "eternal_flame",
        displayName = "Eternal Flame",
        description = "An ember carried between worlds. Every future Earth starts with its fires already burning: +2 Natural Fire per level.",
        icon = "🔥",
        baseCost = 50.0,
        costGrowth = 1.8,
        maxLevel = 10,
        effect = PrestigeUpgradeEffect(startingGenerators = mapOf("natural_fire" to 2)),
    ),
    PrestigeUpgrade(
        id = "extended_endurance",
        displayName = "Extended Endurance",
        description = "Your civilizations keep running longer without you watching. +100% offline progress cap per level.",
        icon = "⏳",
        baseCost = 5_000.0,
        costGrowth = 3.5,
        maxLevel = 3,
        effect = PrestigeUpgradeEffect(offlineCapMultiplier = 2.0),
    ),
    PrestigeUpgrade(
        id = "head_start",
        displayName = "Head Start",
        description = "Begin each Earth with a stockpile of Energy, skipping the slowest opening minutes.",
        icon = "🚩",
        baseCost = 20_000.0,
        costGrowth = 1.0,
        maxLevel = 1,
        effect = PrestigeUpgradeEffect(startingResources = mapOf(ResourceId.ENERGY to 1000.0)),
    ),
    PrestigeUpgrade(
        id = "automated_industry",
        displayName = "Automated Industry",
        description = "Robots keep the lines running while you sleep. +100% offline progress cap, and every future Earth starts with 25 Natural Fires already lit.",
        icon = "🤖",
        baseCost = 200_000.0,
        costGrowth = 1.0,
        maxLevel = 1,
        effect = PrestigeUpgradeEffect(
            offlineCapMultiplier = 2.0,
            startingGenerators = mapOf("natural_fire" to 25),
        ),
    ),
    PrestigeUpgrade(
        id = "deep_foundations",
        displayName = "Deep Foundations",
        description = "Start every future Earth with a Research head start, so the tree opens immediately instead of after the first hour.",
        icon = "🧱",
        baseCost = 40_000.0,
        costGrowth = 1.0,
        maxLevel = 1,
        effect = PrestigeUpgradeEffect(
            startingResources = mapOf(ResourceId.RESEARCH to 5_000.0, ResourceId.ENERGY to 25_000.0),
        ),
    ),
)

val PRESTIGE_UPGRADE_BY_ID: Map<String, PrestigeUpgrade> = PRESTIGE_UPGRADES.associateBy { it.id }

fun prestigeUpgradeCost(upgrade: PrestigeUpgrade, currentLevel: Int): GameDecimal =
    gd(upgrade.baseCost) * gd(upgrade.costGrowth).pow(currentLevel)

/**
 * Everything owned prestige upgrades and completed challenges do, aggregated
 * once so the per-tick production maths is a single lookup.
 */
data class PrestigeMultipliers(
    val global: Double = 1.0,
    val research: Double = 1.0,
    val allGas: Double = 1.0,
    val perGas: Map<GasId, Double> = emptyMap(),
    val techCostDiscount: Double = 0.0,
    val offlineCapMultiplier: Double = 1.0,
    val startingTechIds: List<String> = emptyList(),
    val startingGenerators: Map<String, Int> = emptyMap(),
    val startingResources: Map<ResourceId, Double> = emptyMap(),
) {
    companion object {
        val NONE = PrestigeMultipliers()
    }
}

private class PrestigeAccumulator {
    var global = 1.0
    var research = 1.0
    var allGas = 1.0
    val perGas = mutableMapOf<GasId, Double>()
    var techCostDiscount = 0.0
    var offlineCapMultiplier = 1.0
    val startingTechIds = mutableListOf<String>()
    val startingGenerators = mutableMapOf<String, Int>()
    val startingResources = mutableMapOf<ResourceId, Double>()

    fun apply(e: PrestigeUpgradeEffect, level: Int) {
        e.globalProductionMultiplier?.let { global *= it.pow(level) }
        e.researchMultiplier?.let { research *= it.pow(level) }
        e.allGasProductionMultiplier?.let { allGas *= it.pow(level) }
        e.gasProductionMultiplier?.let { (gas, multiplier) ->
            perGas[gas] = (perGas[gas] ?: 1.0) * multiplier.pow(level)
        }
        e.techCostDiscount?.let { techCostDiscount = min(0.9, techCostDiscount + it * level) }
        e.offlineCapMultiplier?.let { offlineCapMultiplier *= it.pow(level) }
        startingTechIds += e.grantsStartingTechIds
        for ((techId, count) in e.startingGenerators) {
            startingGenerators[techId] = (startingGenerators[techId] ?: 0) + count * level
        }
        for ((resource, amount) in e.startingResources) {
            startingResources[resource] = (startingResources[resource] ?: 0.0) + amount
        }
    }

    fun build() = PrestigeMultipliers(
        global = global,
        research = research,
        allGas = allGas,
        perGas = perGas.toMap(),
        techCostDiscount = techCostDiscount,
        offlineCapMultiplier = offlineCapMultiplier,
        startingTechIds = startingTechIds.toList(),
        startingGenerators = startingGenerators.toMap(),
        startingResources = startingResources.toMap(),
    )
}

/**
 * Aggregates every owned prestige upgrade at every level, plus any permanent
 * challenge rewards, into one multiplier bundle.
 */
fun computePrestigeMultipliers(
    upgradesOwned: Map<String, Int>,
    challengeRewardEffects: List<PrestigeUpgradeEffect> = emptyList(),
): PrestigeMultipliers {
    val accumulator = PrestigeAccumulator()

    for (upgrade in PRESTIGE_UPGRADES) {
        val level = upgradesOwned[upgrade.id] ?: 0
        if (level <= 0) continue
        accumulator.apply(upgrade.effect, level)
    }

    for (reward in challengeRewardEffects) {
        accumulator.apply(reward, 1)
    }

    return accumulator.build()
}

data class PrestigeGainParams(
    val totalGasProducedKg: GasAmounts,
    val peakForcingWm2: Double,
    val civLevel: Int,
    val runDurationSeconds: Double,
)

/**
 * How much faster than the reference pace this run was, as a multiplier on the
 * payout. This is the term that makes each Earth worth more than the last:
 * prestige upgrades buy speed, and speed is what is scored.
 *
 * A run of zero (or nonsensical) length is treated as the reference pace rather
 * than as infinitely fast, so a corrupted clock cannot mint points.
 */
fun speedMultiplier(runDurationSeconds: Double): Double {
    if (!runDurationSeconds.isFinite() || runDurationSeconds <= 0) return 1.0
    val ratio = PRESTIGE.referenceRunSeconds / runDurationSeconds
    val scaled = ratio.pow(PRESTIGE.speedExponent)
    return min(PRESTIGE.maxSpeedMultiplier, max(PRESTIGE.minSpeedMultiplier, scaled))
}

/**
 * Converts one run's outcome into Earth Points: how much gas the civilization
 * produced, scaled by how hard it pushed the atmosphere, how far up the tech
 * tree it got, and how quickly it managed all three.
 */
fun calculatePrestigeGain(params: PrestigeGainParams): GameDecimal {
    val totalGasKg = params.totalGasProducedKg.sum()
    val gasScore = (totalGasKg / PRESTIGE.baseDivisor).clampMin(GameDecimal.ZERO)
    val gasComponent = gasScore.pow(PRESTIGE.exponent) * PRESTIGE.gasWeight

    val forcingMultiplier = 1.0 + PRESTIGE.forcingWeight * sqrt(max(0.0, params.peakForcingWm2))
    val civMultiplier = 1.0 + PRESTIGE.civLevelWeight * sqrt(max(0.0, params.civLevel.toDouble()))

    return gasComponent * forcingMultiplier * civMultiplier * speedMultiplier(params.runDurationSeconds)
}
